src/ngraph/pass/concat_fusion.cpp

```diff
@@ -0,0 +1,304 @@
+//*****************************************************************************
+// Copyright 2017-2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include "concat_fusion.hpp"
+#include <algorithm>
+#include <iostream>
+#include <numeric>
+#include <unordered_set>
+
+#include "ngraph/graph_util.hpp"
+#include "ngraph/log.hpp"
+#include "ngraph/op/broadcast.hpp"
+#include "ngraph/op/concat.hpp"
+#include "ngraph/op/parameter.hpp"
+#include "ngraph/op/reshape.hpp"
+#include "ngraph/pattern/matcher.hpp"
+#include "ngraph/pattern/op/label.hpp"
+#include "ngraph/pattern/op/skip.hpp"
+#include "ngraph/util.hpp"
+
+using namespace ngraph;
+
+bool check_self_concat_op(const std::shared_ptr<Node>& op)
+{
+    auto input_args = op->get_arguments();
+    std::set<std::shared_ptr<Node>> input_args_set(input_args.begin(), input_args.end());
```

Comment:
The comparison of `std::shared_ptr<Node>` is kind of tricky. Use raw pointer Node* or node name is more readable.

Reply:
The idea here is that a self concat will have multiple instances of the same node in `input_args`. Creating a set out of that vector is a quick way of determining whether it is a self concat since the set will have just one instance of the node. Can you please elaborate which comparison you are referring to? Sorry I am a little confused.

Reply:
I mean `std::set<Node*>` will be more straight forward, as we know what to compare. For `std::set<std::shared_ptr<Node>>` I need to check the manual to make sure that `shared_ptr` is compared by the raw pointer it wrapped.
But I understand the implementation will be less elegant. No a big issue though.